src/server/journal/streamer.cc

```diff
@@ -11,62 +11,174 @@
 
 namespace dfly {
 using namespace util;
+using namespace journal;
+using namespace facade;
 
-void JournalStreamer::Start(io::Sink* dest, bool send_lsn) {
-  using namespace journal;
-  write_fb_ = fb2::Fiber("journal_stream", &JournalStreamer::WriterFb, this, dest);
+namespace {
+
+iovec IoVec(io::Bytes src) {
+  return iovec{const_cast<uint8_t*>(src.data()), src.size()};
+}
+
+constexpr chrono::steady_clock::duration kSinkThrottleTime = 500ms;
+
+}  // namespace
+
+JournalStreamer::JournalStreamer(journal::Journal* journal, Context* cntx)
+    : journal_(journal), cntx_(cntx) {
+}
+
+JournalStreamer::~JournalStreamer() {
+  DCHECK_EQ(in_flight_bytes_, 0u);
+  VLOG(1) << "~JournalStreamer";
+}
+
+void JournalStreamer::Start(io::AsyncSink* dest, bool send_lsn) {
+  CHECK(dest_ == nullptr && dest != nullptr);
+  dest_ = dest;
   journal_cb_id_ =
       journal_->RegisterOnChange([this, send_lsn](const JournalItem& item, bool allow_await) {
-        if (!ShouldWrite(item)) {
-          return;
+        if (allow_await) {
+          ThrottleIfNeeded();
+          // No record to write, just await if data was written so consumer will read the data.
+          if (item.opcode == Op::NOOP)
+            return;
         }
 
-        if (item.opcode == Op::NOOP) {
-          // No record to write, just await if data was written so consumer will read the data.
-          return AwaitIfWritten();
+        if (!ShouldWrite(item)) {
+          return;
         }
 
-        Write(io::Buffer(item.data));
+        Write(item.data);
         time_t now = time(nullptr);
+
+        // TODO: to chain it to the previous Write call.
         if (send_lsn && now - last_lsn_time_ > 3) {
           last_lsn_time_ = now;
-          base::IoBuf tmp;
-          io::BufSink sink(&tmp);
+          io::StringSink sink;
           JournalWriter writer(&sink);
           writer.Write(Entry{journal::Op::LSN, item.lsn});
-          Write(io::Buffer(io::View(tmp.InputBuffer())));
+          Write(sink.str());
         }
-        NotifyWritten(allow_await);
       });
 }
 
 void JournalStreamer::Cancel() {
-  Finalize();  // Finalize must be called before UnregisterOnChange because we first need to stop
-               // writing to buffer and notify the all the producers.
-               // Writing to journal holds mutex protecting change_cb_arr_, than the fiber can
-               // preemt when calling NotifyWritten and it will not run again till notified.
-               // UnregisterOnChange will try to lock the mutex therefor calling UnregisterOnChange
-               // before Finalize may cause deadlock.
+  VLOG(1) << "JournalStreamer::Cancel";
+  cntx_->Cancel();
```

Comment:
should I call `cntx_->Cancel();` here @dranikpg  ?

Reply:
I assume not, becauase it cancel is already called when cleaning up. I'd remove it just to be sure, but the error should be ignored if the context already has one set